### ISSUE TITLE
java.lang.IllegalArgumentException: Internal error: Failed to apply delegate: Attempting to use a delegate that only supports static-sized tensors with a graph that has dynamic-sized tensors

### ISSUE THREAD

**Issue by sirius0503** [tensorflow/tensorflow#38036]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information** 
- Have I written custom code (as opposed to using a stock
example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g.,
Linux Ubuntu 16.04): Linux Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if
the issue happens on mobile device: Nokia 7.1
- TensorFlow installed from (source or
binary): - TensorFlow version (use command below): tf-nightly 2.2-dev
- Python version: - Bazel 3.7
version (if compiling from source):
- GCC/Compiler version (if compiling from
source): 
- CUDA/cuDNN version: - GPU model and memory: Mobile gpu Adreno 506

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**

I am trying to use the following [model code here](https://stackoverflow.com/a/60869999/8030107) and have made a tensorflow lite model, when we run it on android we get the following error using the GPU delegate:

```
java.lang.IllegalArgumentException: Internal error: Failed to apply delegate: Attempting to use a delegate that only supports static-sized tensors with a graph that has dynamic-sized tensors.
```

And when we switch to the CPU version we get the following error:

```

 Cannot convert between a TensorFlowLite tensor with type FLOAT32 and a Java object of type [I (which is compatible with the TensorFlowLite type INT32).
------------------------------------------------
whole error -> 2020-03-30 12:26:15.613 3471-3471/com.valuepitch.intruderdetector W/System.err: java.lang.IllegalArgumentException: Cannot convert between a TensorFlowLite tensor with type FLOAT32 and a Java object of type [I (which is compatible with the TensorFlowLite type INT32).
2020-03-30 12:26:15.614 3471-3471/com.valuepitch.intruderdetector W/System.err:     at org.tensorflow.lite.Tensor.throwIfTypeIsIncompatible(Tensor.java:316)
2020-03-30 12:26:15.614 3471-3471/com.valuepitch.intruderdetector W/System.err:     at org.tensorflow.lite.Tensor.getInputShapeIfDifferent(Tensor.java:218)
2020-03-30 12:26:15.614 3471-3471/com.valuepitch.intruderdetector W/System.err:     at org.tensorflow.lite.NativeInterpreterWrapper.run(NativeInterpreterWrapper.java:137)
2020-03-30 12:26:15.614 3471-3471/com.valuepitch.intruderdetector W/System.err:     at org.tensorflow.lite.Interpreter.runForMultipleInputsOutputs(Interpreter.java:311)
```

**Describe the expected behavior**

I expect the model to take 3 arguments and ouput the model outputs. I've seen similar issue [here](https://github.com/tensorflow/tensorflow/issues/25131) , but couldn't comment there for unknown reasons or might be because the issue was closed? No clue as to that.

**Standalone code to reproduce the issue** 
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by jvishnuvardhan** [comment]
@sirius0503 Can you please share a colab? Current link not pointing to colab. Thanks!
Your issue looks more like [This](https://github.com/tensorflow/tensorflow/issues/37706#issuecomment-601358807) issue. Please take a look at the responses. Thanks!

**Comment by sirius0503** [comment]
> @sirius0503 Can you please share a colab? Current link not pointing to colab. Thanks!
> Your issue looks more like [This](https://github.com/tensorflow/tensorflow/issues/37706#issuecomment-601358807) issue. Please take a look at the responses. Thanks!

Here's the [colab link](https://colab.research.google.com/drive/1sPfULw3TDjWUvHvAaMXweBxwWeGfj25j), have updated it in the question too!

**Comment by karimnosseir** [comment]
As the error message shows, your model has dynamic shapes and the GPU delegate doesn't support this.
You will need to fully specify the input shape (Static sizes) during conversion, and then the model can be used with gpu delegate.

**Comment by sirius0503** [comment]
> As the error message shows, your model has dynamic shapes and the GPU delegate doesn't support this.
> You will need to fully specify the input shape (Static sizes) during conversion, and then the model can be used with gpu delegate.

@karimnosseir When I use `batch_size=None` and pass a `[1000, 4]` shape, then it works with the **cpu**, but produces this error for the `gpu` and `nnapi` delegate, any reason why?

I have asked the same thing [here](https://github.com/tensorflow/tensorflow/issues/37706#issuecomment-606507426), @jdduke had asked me for a tflite model there.

**Comment by jdduke** [comment]
We're working in improving support for this. Note that, even if we do get this working correctly in the short term (resizing graphs when using GPU/NNAPI acceleration), it's unlikely that you'll want to be resizing with any frequency, as there is a fixed cost for recreating the graph and regenerating associated shaders.

**Comment by sirius0503** [comment]
> We're working in improving support for this. Note that, even if we do get this working correctly in the short term (resizing graphs when using GPU/NNAPI acceleration), it's unlikely that you'll want to be resizing with any frequency, as there is a fixed cost for recreating the graph and regenerating associated shaders.

@jdduke Can you explain the resizing graphs with any frequency and the fixed cost and the associated shaders part part in a more simple manner and/or provide some links where I can read up more on them. Thanks a lot!

Also, where can I follow the development/support for this for the GPU/NNAPI acceleration?

**Comment by jdduke** [comment]
Hey @sirius0503 , apologies for the delay.

> Can you explain the resizing graphs with any frequency and the fixed cost and the associated shaders part part in a more simple manner and/or provide some links where I can read up more on them. 

Sure! The short answer is that, when you use GPU delegation (or NNAPI delegation), there is a preparation stage where the graph is effectively recompiled into a different format, based on the current tensor shapes in the graph. For the GPU, this means creating and compiling a sequences of GPU shaders. For NNAPi, this means creating an NNAPI model, underneath which there can be additional preparation/processing done by various NNAPi drivers. All of this preparation can be expensive in terms of latency, often more expensive than a single inference call.

Currently, when an input is resized, all of that preparation work has to be executed *again* based on the new graph shapes. In an ideal world, the hardware backends could accommodate dynamic shapes, avoiding the need to "recompile" the model after a resize. However, implementing this behavior is complicated.

I can't give a concrete timeline as to when this will be possible. Some workarounds that clients have used:
 * Create several different interpreters with various batch sizes, and use the right one depending on the current batch input.
 * Use a larger batch size than may be needed, always running with that batch size, but only feed in a smaller number of batches if necessary during inference.

**Comment by tastelessjolt** [comment]
I'm having a similar problem. 
```
Internal error: Failed to run on the given Interpreter: Attempting to use a delegate that only supports static-sized tensors with a graph that has dynamic-sized tensors.
```

Though I've used static shapes like input1: (1, 256, 256, 3) and input2: (1, 256, 256, 3) while converting the model (using concrete function) to TFLite using both [tf.lite.OpsSet.TFLITE_BUILTINS, tf.lite.OpsSet.SELECT_TF_OPS] ops. 

So, **Question:** How do you check whether the TFlite file has static sizes or dynamic sizes and which of the nodes/tensors have such dynamic sizes?

Misc Info:
Using Tensorflow version 2.2 binary from Anaconda

PS: Can't post the TFlite model due to company restrictions.

Thanks in advance!

**Comment by karimnosseir** [comment]
@tastelessjolt Does your model have an op that produce dynamic shape ?
Try visualizing it and inspect which op causes this.
If you have a reproduce example, that will be good to check what is your problem.

Thanks

**Comment by tastelessjolt** [comment]
@karimnosseir can you explain what do mean by a reproduce example? I don't understand.  

I'll try visualizing it and check for ops producing dynamic shape. But how is it different? Since you know the shape of input of the network, doesn't it mean you know the shape of all the tensors in the graph? 

Btw, can you also suggest some visualizing tool for tflite? I use Netron, it seems fine but if there's a better one out there I'd like that. 

Thanks

**Comment by karimnosseir** [comment]
@tastelessjolt a model that trigger this case when specifying static shape in conversion.

You can use Netron

**Comment by tastelessjolt** [comment]
Does `TransposeConv` output a dynamic shape? Also in Netron, as far as I can see (The model is quite big) all the shapes are defined. Also, all the outputs of the model have outputs defined.

**Comment by tastelessjolt** [comment]
Also, does reduce_mean and reduce_sum and similar aggregating operators produce dynamic shapes?
Because I do have a few of them.

**Comment by tastelessjolt** [comment]
@karimnosseir Can you please reply to some of my doubts above? I'm tagging you as I forgot to tag you in the previous comments. 

Thank you

**Comment by karimnosseir** [comment]
@tastelessjolt It depends on your model. Please share reproduce example, so we can have a look.

Thanks

**Comment by InfiniteLife** [comment]
Hi! I encountered same error. Model runs on Android-Cpu fine, but on Gpu gives the same error.
I converted model using [1,512,512,3] shapes, so inputs are static, but I think problem is in the outputs because Neutron does not shows their shapes:

![image](https://user-images.githubusercontent.com/7304884/85278019-04138d00-b48d-11ea-9fa9-62ecf62c74c0.png)

The model is efficientdet-d0 from https://github.com/google/automl, I think I will be able to upload model a little bit later.

**Comment by pranshugo** [comment]
I am having the same issue and I am pretty sure that my graph does not have any dynamic shapes because I looked that up and fixed the batch size while training the data. But adding tflite nnapi delegates still say that "Attempting to use a delegate that only supports static-sized tensors with a graph that has dynamic-sized tensors"

**Comment by pranshugo** [comment]
yeah alright, here is my tflite model

https://drive.google.com/file/d/10COhJsGLiu15y51IyVAd2eoz8me8iixn/view?usp=sharing 
adding the drive link as the github is not allowing files larger than 10MB

**Comment by ryanaleksander** [comment]
Having the same problem, here is my model

[model_new.zip](https://github.com/tensorflow/tensorflow/files/5122767/model_new.zip)

**Comment by pranshugo** [comment]
Any news? @jdduke @karimnosseir

**Comment by jdduke** [comment]
@srjoglekar246 can help diagnose the issue in the model from [your earlier comment](https://github.com/tensorflow/tensorflow/issues/38036#issuecomment-678924619), and why it has dynamic shapes.

**Comment by karimnosseir** [comment]
Sorry for the late reply.

@pranshugo Your model has TF ops (converted with SELECT_TF) currently this introduces dynamic shape in the runtime. It's on our radar to fix SELECT option to avoid dynamic shape during inference.

**Comment by karimnosseir** [comment]
@ryanaleksander Your model has dynamic shape, but i think we can get the information during conversion. Do you mind sharing your original TF graph, i want to check why the shape is missing during conversion.

Thanks

**Comment by Lizhi-Liao** [comment]
[lstm_tflite_model.zip](https://github.com/tensorflow/tensorflow/files/5166095/lstm_tflite_model.zip)
Having the same problem and here is my model
Thanks a lot! 
@karimnosseir

**Comment by ryanaleksander** [comment]
@karimnosseir Thanks for looking into my problem, here is the original TF model
[frozen_inference_graph.zip](https://github.com/tensorflow/tensorflow/files/5166146/frozen_inference_graph.zip)

**Comment by pranshugo** [comment]
Hey @karimnosseir, can you suggest some method on how can I convert my graph to tflite without using SELECT_TF_OPS tflite supported ops flag. Because without it, the converter raises an exception. Or what could be in my graph that needs SELECT_TF_OPS?

**Comment by srjoglekar246** [comment]
@pranshugo Can you post your model with SELECT_TF_OPS? We can identify which TF ops are required to be made into TFLite custom ops (so that we dont need SELECT_TF_OPS)

**Comment by pranshugo** [comment]
hey @srjoglekar246, 
Here is my model converted with SELECT_TF_OPS flag: https://drive.google.com/file/d/10COhJsGLiu15y51IyVAd2eoz8me8iixn/view?usp=sharing

Also, I am attaching my frozen graph along, so if you can suggest me a method to convert it with static shapes.
I am using tensorflow2.2.0
https://drive.google.com/file/d/16j7w8KwUnaxHvygsAygk5DM1qK61x4cs/view?usp=sharing

**Comment by monadiconion** [comment]
I got the same **"Internal error: Failed to apply delegate: Attempting to use a delegate that only supports static-sized tensors with a graph that has dynamic-sized tensors"** error running CREPE model on TF Lite (https://github.com/marl/crepe).

Here is the converted model (from the original Keras model):
https://drive.google.com/file/d/16IlycajzJqo5NtXv8N_7PqE3QfDUUDRd/view?usp=sharing

Keras model summary
```
Model: "functional_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input (InputLayer)           [(None, 1024)]            0         
_________________________________________________________________
input-reshape (Reshape)      (None, 1024, 1, 1)        0         
_________________________________________________________________
conv1 (Conv2D)               (None, 256, 1, 128)       65664     
_________________________________________________________________
conv1-BN (BatchNormalization (None, 256, 1, 128)       512       
_________________________________________________________________
conv1-maxpool (MaxPooling2D) (None, 128, 1, 128)       0         
_________________________________________________________________
conv1-dropout (Dropout)      (None, 128, 1, 128)       0         
_________________________________________________________________
conv2 (Conv2D)               (None, 128, 1, 16)        131088    
_________________________________________________________________
conv2-BN (BatchNormalization (None, 128, 1, 16)        64        
_________________________________________________________________
conv2-maxpool (MaxPooling2D) (None, 64, 1, 16)         0         
_________________________________________________________________
conv2-dropout (Dropout)      (None, 64, 1, 16)         0         
_________________________________________________________________
conv3 (Conv2D)               (None, 64, 1, 16)         16400     
_________________________________________________________________
conv3-BN (BatchNormalization (None, 64, 1, 16)         64        
_________________________________________________________________
conv3-maxpool (MaxPooling2D) (None, 32, 1, 16)         0         
_________________________________________________________________
conv3-dropout (Dropout)      (None, 32, 1, 16)         0         
_________________________________________________________________
conv4 (Conv2D)               (None, 32, 1, 16)         16400     
_________________________________________________________________
conv4-BN (BatchNormalization (None, 32, 1, 16)         64        
_________________________________________________________________
conv4-maxpool (MaxPooling2D) (None, 16, 1, 16)         0         
_________________________________________________________________
conv4-dropout (Dropout)      (None, 16, 1, 16)         0         
_________________________________________________________________
conv5 (Conv2D)               (None, 16, 1, 32)         32800     
_________________________________________________________________
conv5-BN (BatchNormalization (None, 16, 1, 32)         128       
_________________________________________________________________
conv5-maxpool (MaxPooling2D) (None, 8, 1, 32)          0         
_________________________________________________________________
conv5-dropout (Dropout)      (None, 8, 1, 32)          0         
_________________________________________________________________
conv6 (Conv2D)               (None, 8, 1, 64)          131136    
_________________________________________________________________
conv6-BN (BatchNormalization (None, 8, 1, 64)          256       
_________________________________________________________________
conv6-maxpool (MaxPooling2D) (None, 4, 1, 64)          0         
_________________________________________________________________
conv6-dropout (Dropout)      (None, 4, 1, 64)          0         
_________________________________________________________________
transpose (Permute)          (None, 1, 4, 64)          0         
_________________________________________________________________
flatten (Flatten)            (None, 256)               0         
_________________________________________________________________
classifier (Dense)           (None, 360)               92520     
=================================================================
Total params: 487,096
Trainable params: 486,552
Non-trainable params: 544
_________________________________________________________________
```

I am using TF 2.3.0 and here is the code for the model conversion
```
converter = tf.lite.TFLiteConverter.from_keras_model(model)
converter.experimental_new_converter = True
converter.optimizations = [tf.lite.Optimize.DEFAULT]
tflite_model = converter.convert()
```

From https://github.com/tensorflow/tensorflow/issues/24607, it seems that the dynamic size issue has been fixed with `converter.experimental_new_converter = True` (I guess the shape with "None" will be converted to 1?). So is there any other "dynamic-sized tensors / operators" included in this model causing the issue and anything I could do with it?

**Comment by srjoglekar246** [comment]
@monadiconion Your model's input shape has an 'unknown' batch dim (The `None` above). If you can build the Keras model such that it has a fixed input shape (maybe by setting batch=1), the model will get acceleration support. Delegate backends work with fixed-shape graphs *only*, to avoid constant re-compilation of internal data structures.

@pranshugo Which version of TF/tflite_convert are you using for your conversion? Your model contains a ResizeNearestNeighbor op, which is causing the issue. I think its being used in a way not supported by TFLite. Can you post the snippet of code where its coming from?

**Comment by pranshugo** [comment]
I am using TF 2.2.0: 
$pip3 show tensorflow-gpu
```
Name: tensorflow-gpu
Version: 2.2.0
Summary: TensorFlow is an open source machine learning framework for everyone.
Home-page: https://www.tensorflow.org/
Author: Google Inc.
Author-email: packages@tensorflow.org
License: Apache 2.0
Location: /home/nymble/.local/lib/python3.6/site-packages
Requires: absl-py, grpcio, h5py, astunparse, keras-preprocessing, gast, numpy, protobuf, six, termcolor, tensorboard, google-pasta, wheel, tensorflow-estimator, opt-einsum, scipy, wrapt
Required-by: 
```

I use this code in converting my model
`converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS, 
                                            tf.lite.OpsSet.SELECT_TF_OPS]`

In my model, I use keras.layers.Upsampling2D layer in Keras, which in turn gets converted to ResizeNearestNeighbour operation in tflite.
Converting my model without above converter supported ops gives error as follows:
`<unknown>:0: error: failed while converting: 'main': Ops that can be supported by the flex runtime (enabled via setting the -emit-select-tf-ops flag): ResizeNearestNeighbor,ResizeNearestNeighbor,ResizeNearestNeighbor.`

Is upsampling2D layer not supported by tflite? or is there anything I can use in place of upsampling2D?

**Comment by zldrobit** [comment]
@pranshugo maybe you could try to use `tf.image.resize` instead of `keras.layers.Upsampling2D`.
In TF2.3, I encountered a same problem which I solved by using the above method.

**Comment by pranshugo** [comment]
That worked out, 
Thanks @zldrobit 
Also I found out that tf.keras.Layers.Upsampling2D with interpolation argument as 'nearest' isn't supported by tflite, whereas 'bilnear' works.

**Comment by jdduke** [comment]
> Also I found out that tf.keras.Layers.Upsampling2D with interpolation argument as 'nearest' isn't supported by tflite, whereas 'bilnear' works.

Which version of TensorFlow are you using? Can you file a separate bug? We do support ResizeNearestNeighbor in TFLite, but more recently added support for some attributes (e.g., half_pixel_centers).

**Comment by pranshugo** [comment]
#43481 
@jdduke Here, I filed a separate bug regarding this.

**Comment by jangmin-deepx** [comment]
@pranshugo Official Release version of tflite is 2.1.0 right now and it doesn't support ResizeNearestNeighbor with unsupported problem('RESIZE_NEAREST_NEIGHBOR' version '3' for custom YOLO)

I found [ https://github.com/PINTO0309/TensorflowLite-bin ] which supports 2.3.0 tflite-runtime and it fix the prev error.

**Comment by jangmin-deepx** [comment]
I was getting same Problem and i solved !!

Try This code Below
env :
TF :2.3
TF-Lite-Runtime : 2.3

The Major point was Matching OPs betweene. TF vs TFLite
```python


c = tf.lite.TFLiteConverter.from_keras_model(yolo)
c.target_spec.supported_ops =[tf.lite.OpsSet.TFLITE_BUILTINS, tf.lite.OpsSet.SELECT_TF_OPS]
tflite_model = c.convert()
open("./v3_tiny_noop.tflite", 'wb').write(tflite_model)

```

I saw some opinion that "experimental_new_converter" attr sometime ocurring problems.
But in my case  ( TF 2.3, TF-L-Runtime 2.3 ) none-toco-ver fixesd problem.

Custum op problem, static-sized tensors Problem

I hope anyone suffering from this prob getting relieved

**Comment by le8888e** [comment]
Hi @Lizhi-Liao ,

I met the same problem with me. In my case there's a BiLSTM layer in my model and I implement it by Keras.
Have you fix it？

Thank you.

**Comment by le8888e** [comment]
Hi @srjoglekar246 @jdduke @karimnosseir ,

I came with the same problem：
'RuntimeError: Attempting to use a delegate that only supports static-sized tensors with a graph that has dynamic-sized tensors. Failed to apply the default TensorFlow Lite delegate indexed at 0.'
 I'm trying to run a tflite model on x86 CPU with XNNPack complied from TF2.4 source code. Here is my model:
https://drive.google.com/file/d/1QZHUAao4rMo64Oy9KhdD_X1-0aAy32u2/view?usp=sharing
There is a BiLSTM layer declared by Keras otherwise I cannot convert from .pb to .tflite. Other layers are declared by TF.
I use 'converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS, tf.lite.OpsSet.SELECT_TF_OPS]' for converting, otherwise there would be a bug.
By the way when I run inference on this model with tf-nightly 2.4 built by pip, no error occurs. And I can run ResNet50 tflite model with XNNPack complied. Any help or suggestions are appreciated.

Thank you

**Comment by srjoglekar246** [comment]
We only recently added support for SELECT_TF_OPS to work with XNNPack. So this should only work with the latest nightly, and may not work with the last stable TF release.

**Comment by le8888e** [comment]
@srjoglekar246 So if I install the latest nightly via pip, will XNNPack be complied by default?
Thank you

**Comment by srjoglekar246** [comment]
XNNPack will work with SELECT_TF_OPS without any error at runtime.

**Comment by le8888e** [comment]
@srjoglekar246 As I know, XNNPack can only be complied by building  tensorflow from source code. So how to enable XNNPack in the lastest tf-nightly? Thank you.

**Comment by hamlatzis** [comment]
You cannot always know beforhand the output shape of a tensor. What about the unique op? Both in TF and TFLite the shape of the first output tensor is calculated after everything is evaluated. 

So in OpenCL how do you implement this? When you only know it's a 1D tensor?

**Comment by objectc** [comment]
> @monadiconion Your model's input shape has an 'unknown' batch dim (The `None` above). If you can build the Keras model such that it has a fixed input shape (maybe by setting batch=1), the model will get acceleration support. Delegate backends work with fixed-shape graphs _only_, to avoid constant re-compilation of internal data structures.

@srjoglekar246 
Almost all the pre-trained models have such kind of placeholder for batch dim, and some may even have other dynamic dims like width and height. In fact, most of the immediate layer's output shape are dependent on the input size. Is there any way to set all the layers with a corresponding fixed value other than rebuild the model from scratch? It seems no API in TensorFlow can do it directly. How did you solve it when generated hosted models from pre-trained models in [TensorFlow Hub](https://tfhub.dev/s?deployment-format=lite)? Thanks.

**Comment by srjoglekar246** [comment]
@hamlatzis @objectc  You do not need to specify the shape of every intermediate tensor, since TFLite propagate shapes from the input dimensions. You only need to set the *input* shapes. Also note that you can resize the batch dims later, but you need something other than 'unknown' for on-device accelerators to fuse operations together and build their internal data structures. See [this](http://google3/third_party/tensorflow_models/object_detection/export_tflite_graph_lib_tf2.py?l=245&rcl=329938888) as an example of how to set input shape while generating a SavedModel for TFLite conversion.

**Comment by hamlatzis** [comment]
@srjoglekar246 

regarding my comment (https://github.com/tensorflow/tensorflow/issues/38036#issuecomment-702746904) I was referring to OpenCL code that runs on the GPU. 

For the CPU both tensorflow and tflite for the unique op (https://www.tensorflow.org/api_docs/python/tf/unique) the first tensor is 1D but shape is calculated dynamically when the unique values are found. Let's say now that you want to make an implementation of your own to run on the GPU. Your options are OpenCL, OpenGL and Metal delegates (depending on the OS). I am interested in OpenCL. As far as I know you cannot have dynamic arrays in OpenCL, so if I were to implement this op in OpenCL I can think of only two ways.

1. Allocate a static buffer with the maximum possible size, as if all values are unique. But it's a waste of memory in case I have all values the same. In this case TF and tflite just allocate a tensor of size 1

2. Run the code on CPU and calculate the unique values, pass the size of the tensor to OpenCL and then run the op there. But this is a waste of processing power, I would have run the op twice, once on the CPU and once on the GPU

I assume similar problems exist for other ops that the shape is calculated at runtime.

**Comment by srjoglekar246** [comment]
Which is why our GPU backend doesn't support the case where shapes are generated at runtime :-)

**Comment by nadir121** [comment]
> Which is why our GPU backend doesn't support the case where shapes are generated at runtime :-)

So is it never going to be addressed?

**Comment by srjoglekar246** [comment]
> So is it never going to be addressed?

That is not the case. It hasn't been prioritized yet, but we plan to look at it in the future.

**Comment by objectc** [comment]
> @hamlatzis @objectc You do not need to specify the shape of every intermediate tensor, since TFLite propagate shapes from the input dimensions. You only need to set the _input_ shapes. Also note that you can resize the batch dims later, but you need something other than 'unknown' for on-device accelerators to fuse operations together and build their internal data structures. See [this](https://github.com/tensorflow/models/blob/master/research/object_detection/export_tflite_graph_lib_tf2.py#L245) as an example of how to set input shape while generating a SavedModel for TFLite conversion.

@srjoglekar246 

This works for part of the models. Thanks.

What if the model contains a [custom layer](https://github.com/fizyr/keras-retinanet/blob/01737e6523c09df922d36edac332a68bcda4af90/keras_retinanet/layers/filter_detections.py#L54) with operations like `greater`, `pad`, and `gather` that would generate results with the unknown shape? According to my test, even all the sizes of input and output(fixed by `pad`) from each layer are fixed, the same error still there. 

I attached a converted demo model that can reproduce the issue. Any help would be grateful. Thanks! 
[inference_opt_latency.tflite.zip](https://github.com/tensorflow/tensorflow/files/5351115/inference_opt_latency.tflite.zip)

**Comment by srjoglekar246** [comment]
@objectc Did you author the model? Or have you exported/converted it from our model garden? If possible, it could be more useful to explore changes to the source model while we investigate delegate support for graphs with dynamic tensors.

**Comment by objectc** [comment]
> @objectc Did you author the model? Or have you exported/converted it from our model garden? If possible, it could be more useful to explore changes to the source model while we investigate delegate support for graphs with dynamic tensors.

@srjoglekar246 [Here](https://colab.research.google.com/drive/13D5QfXitMeNS8CN3bb5O3SG5mbbnynI8?usp=sharing) is a naive version of it build from scratch. As I mentioned, the operation `tf.greater` would generate output with unknown size. Is it possiable that TF Lite's GPU delegate would support this? Thanks.

**Comment by srjoglekar246** [comment]
It is unlikely, atleast in the short term. We are still exploring use-cases for GPU dynamic-batch support, since it will take quite a bit of investment.

However, for your use-case, may I suggest splitting the model into two TFLite models? This has worked for some clients in the past, where the "heavy" part of the model (Conv, Dense, etc) is accelerated by the GPU, and some post-processing (that typically involves dynamic tensors) is performed in another model. There may be some tradeoff for copying over the first model's outputs to the second's inputs, but the GPU speedup could still be more.

**Comment by objectc** [comment]
Got it, That's exactly what I'm doing right now. Thank you so much for your support!

**Comment by ysyyork** [comment]
> @monadiconion Your model's input shape has an 'unknown' batch dim (The `None` above). If you can build the Keras model such that it has a fixed input shape (maybe by setting batch=1), the model will get acceleration support. Delegate backends work with fixed-shape graphs _only_, to avoid constant re-compilation of internal data structures.
> 
> @pranshugo Which version of TF/tflite_convert are you using for your conversion? Your model contains a ResizeNearestNeighbor op, which is causing the issue. I think its being used in a way not supported by TFLite. Can you post the snippet of code where its coming from?

To add on, this works. if you have keras `Input` layer in the model, make sure you set the batch_size of the Input layer to a fixed value before you convert it to tflite. Otherwise the batch_size dim will remain -1 as dynamic.

**Comment by mohantym** [comment]
Hi @sirius0503 !
We are checking to see if you still need help in this issue .Attaching similar issues  for answers [link1](https://github.com/tensorflow/tensorflow/issues/46388#issuecomment-871074251),[link2](https://github.com/tensorflow/tensorflow/issues/49224),[link3](https://github.com/tensorflow/tensorflow/issues/46388#issuecomment-871074251),[link4](https://stackoverflow.com/questions/60826779/valueerror-none-is-only-supported-in-the-1st-dimension-tensor-flatbuffer-data/60869999#60869999) ? Please create a new issue if the issue is replicating in latest versions. Thanks!

**Comment by Gaozhongpai** [comment]
Hi @karimnosseir, I am having the same problem and here is my model: https://github.com/Gaozhongpai/mediapipe/blob/master/mediapipe/modules/hand_landmark/hand_mesh.tflite Could you help me check it?
